enhancements/network/user-defined-network-segmentation.md

```diff
@@ -0,0 +1,947 @@
+* Kubelet healthchecks/probes will still work on all pods.
+* OpenShift Router/Ingress will work with some limitations for user-defined networks.
+
+### Non-Goals
+
+* Allowing different service CIDRs to be used in different networks.
+* Localnet will not be supported initially for primary networks.
+* Allowing multiple primary networks per namespace.
+* Hybrid overlay support on user-defined networks.
+
+### Future-Goals
+
+* DNS lookup for pods returning records for IPs on the user-defined network. In the first phase DNS will return the pod
+IP on the cluster default network instead.
+* Admin ability to configure networks to have access to all services and/or expose services to be accessible from all
+networks.
+* Ability to advertise user-defined networks to external networks using BGP/EVPN. This will enable things like:
+    * External -> Pod ingress per VRF (Ingress directly to pod IP)
+    * Multiple External Gateway (MEG) in a BGP context, with ECMP routes
+* Allow connection of multiple networks via explicit router API configuration.
+
+## Proposal
+
+By default in OVN-Kubernetes, pods are attached to what is known as the “cluster default" network, which is a routed network
+divided up into a subnet per node. All pods will continue to have an attachment to this network, even when assigned a
+different primary network. Therefore, when a pod is assigned to a user-defined network, it will have two interfaces, one
+to the cluster default network, and one to the user-defined network. The cluster default network is required in order to provide:
+
+1. KAPI service access
+2. DNS service access
+3. Kubelet healthcheck probes to the pod
+4. User defined ports that will be exposed on the cluster default network. This may be used for things like prometheus
+   metric scraping.
+
+All other traffic from the pod will be dropped by firewall rules on this network, when the pod is assigned a user-defined
+primary network. Routes will be added to the pod to route KAPI/DNS traffic out towards the cluster default network. Note,
+it may be desired to allow access to any Kubernetes service on the cluster default network (instead of just KAPI/DNS),
+but at a minimum KAPI/DNS will be accessible. Furthermore, the IP of the pod from the Kubernetes API will continue to
+show the IP assigned in the cluster default network.
+
+In OVN-Kubernetes secondary networks are defined using Network Attachment Definitions (NADs). For more information on
+how these are configured, refer to:
+
+[https://github.com/ovn-org/ovn-kubernetes/blob/master/docs/features/multi-homing.md](https://github.com/ovn-org/ovn-kubernetes/blob/master/docs/features/multi-homing.md)
+
+The proposal here is to leverage this existing mechanism to create the network. The only new field that will be
+introduced to the NAD spec is a “primaryNetwork” key which indicates that this network should be used for the pods
```

Comment:
@trozet I was wondering about a scenario (which I have no clue is real or even if it makes sense) in which a network is primary network on some namespaces, **but** a secondary on others.

Does this make sense ? (I'm not saying it does ...).

Also, I assume we will ensure the value of `primaryNetwork` **must** be equal on *all* NADs granting access to the same network. Am I right here ?

If so, that would prevent the scenario described above.

Reply:
yeah good point,,,

can I have pod1 in ns1 attached to l3-network as user defined primary network
and pod2 in ns2 attached to l2-network as user defined primary network but has l3-network as its secondary network?

Reply:
I think that should be ok, if connecting namespaces together is not a problem.

Reply:
Yeah I think it should be OK, but it may come with some limitations that are not specific to this use case but exist for any secondary network with north/south. With ETP local for example on services, a packet sent to pod1 in @tssurya example, would work and the reply would come back to the l3-network, because the default gateway route goes that way. However, if the ETP local service hit pod2, the reply packet would go towards l2-network cause thats where its default gateway is and most likely get dropped.

ETP=cluster should work fine in both cases because it gets SNAT'ed to join subnet. There may be more cases like this with the introduction of north/south into secondary networks. I think we allow users to add some static routes to the pod dont we? Maybe that would help alleviate any issues where a user wants some north/south traffic go to via a secondary network.

Reply:
yeah @trozet in slack we thought about this more and concluded that we should not be allowing this.
So we need to be explicit here: For a given net-name if multiple NADs refer to the same net-name in conf then the rest of the fields in the conf must be exactly the same; => so its better to not support this case for now.

Reply:
sounds good.